components/forms/LoginForm.jsx

```diff
@@ -10,16 +10,51 @@ class LoginForm extends React.Component {
 
         this.state = {
             username: "",
-            password: ""
-        }
-
+            password: "",
+            isValid: false
+        };
+        this.validityMap = {
+            username: false,
+            password: false
+        };
         this.handleInputChange = this.handleInputChange.bind(this);
+        this.renderField = this.renderField.bind(this);
+        this.validateFields = this.validateFields.bind(this);
     }
 
     handleInputChange(name, value) {
         this.setState({[name]: value});
     }
 
+    validateFields(key, value) {
+        this.validityMap[key] = value;
+        this.setState({
+            isValid: this.validityMap.username && this.validityMap.password
+        });
```

Comment:
Please only set state when there is a change in value. Otherwise, we'll unnecessarily render. IIRC, React is not smart about that.

Reply:
Here I  actually wanted a rerender. The validity has changed so the button needs to rerender as either enabled or disabled. The rerender here will render the `DialogForm` below with the value of `disabledSubmit` updated to the new value of `isValid` set here.

Reply:
Let me clarify with code:

```javascript
this.validityMap[key] = value;
const validity = this.validityMap.username && this.validityMap.password;
if (this.state.isValid !== validity )
        this.setState({
            isValid:validity
        });
```

or better, using the setState updater:
```javascript
this.validityMap[key] = value;
this.setState( (state) => {
    const validity = this.validityMap.username && this.validityMap.password;
    return (state.isValid === validity) ? null //returning null does not set state and prevents render
    : { isValid: validity};  //Please check that this merges state correctly. Otherwise do Object.assign(state, {isValid: validity}); I just forget what works here :-)
});
```
This gist is to not invoke setState unless isValid is changing, or else use the updater that explicitly signals to not update state when isValid is not changing.

Reply:
Ah, I see what you mean now.

In my opinion there are two issues here:

1. We want the state and the logic to be clear and concise
2. We want react to perform optimally.

If I were new to this code, something like 
```
this.setState({
    isValid: this.validityMap.username && this.validityMap.password
});
```
is easier to understand than 
```
this.setState( (state) => {
    const validity = this.validityMap.username && this.validityMap.password;
    return (state.isValid === validity) ? null //returning null does not set state and prevents render
    : { isValid: validity};
});
```

The first version is clearer and more concise in my opinion, but it lacks the performance optimizations of the second. We do have options to keep the shorter setState method while also getting the optimizations:

1. We place the optimization logic inside the setState method by comparing existing state before updating.

I'm just personally not a fan of this because it makes each setState call more complex while spreading the optimization logic across the entire component.

2. We implement the component as extending `PureComponent` instead of `Component`

This would allow the setState method to remain as it is now, but react would do a shallow comparison of the state before re rendering. This shallow comparison would see that the validity did not change and so it would not rerender. Some downsides of this are that it's not clear to everyone why it is or is not triggering a render, and the shallow comparison of state can burn people ocassionally.

3. We implement a `shouldComponentUpdate` method.

This would place all the optimization logic in one place, be clear to the developer when it renders or when it doesn't, and allow the simpler setState, call and give better performance. The problem with this is that when the component is in active development people often forget to add a comparison for a new value inside this function, then have to debug why the component isn't rerendering

The reason that I didn't implement options 2 or 3 from the start is that these performance optimizations have burned me too many times. This may be bad to say but in my experience these optimizations don't give the value they seem. React is pretty fast and there is definitely need to optimize it, but I usually do that towards the end of the dev cycle when the requirements and goal is more concrete. That is unless it is clear that the rendering is computationally expensive and always needs to be optimized.

Reply:
Very good points. I'm actually fairly new to React and appreciate becoming hip to PureComponent. If anything, this conversation is just pushing me to want to move to hooks all that much faster, which should obviate this entirely. Have any thoughts on hooks in this context?

I do find unnecessary re-rendering to be a bear when debugging.

I'm not stuck on taking any particular path here since I want to dump the classes going forward anyways.

Reply:
Hmm, I'll have to think on that one for a bit. I just started using hooks myself and I need to know more about them, specifically how to optimize those. From my understanding of function components with hooks, once you add the `useState` hook and introduce state you can get the same problem that we see here. There is a `useMemo` and `useCallback` that you can add instead to optimize it, but I was reading an article recently saying that those don't give the performance boost that you would expect in all scenarios.

In my opinion the best path forward here might be to just extend `PureComponent` instead of `Component` for now, then I can read/learn a bit more about hooks and we could see what it would look like to start moving some of the classes into function components with hooks.

Reply:
Go for it! PureComponent is great for now.